pgmpy/prediction/DynamicDMLRegressor.py

```diff
@@ -0,0 +1,851 @@
+        self.model_t_ = self._select_model(self.model_t, self.discrete_treatment_)
+
+        # STEP 7: STAGE 1 - Cross-fitted nuisance estimation
+        self._fit_nuisances(X, T, y, groups)
+
+        # STEP 8: STAGE 2 - Backward recursive parameter estimation
+        self._fit_parameters()
+
+        # STEP 9: Compute asymptotic covariance
+        self._compute_covariance()
+
+        # STEP 10: Set sklearn-required attributes
+        self.n_features_in_ = X.shape[1]  # Number of features (columns)
+        self.n_samples_ = X.shape[0]  # Number of samples (rows)
+
+        # STEP 10: Return self
+        return self
+
+    def predict(self, X, T0=None, T1=None):
+        """
+        Predict treatment effects.
+
+        ALGORITHM:
+        Compute E[Y | do(T=T1)] - E[Y | do(T=T0)] using estimated coef_.
+        Effect = sum over periods t of coef_[t] @ (T1_t - T0_t)
+
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
+            Features for prediction.
+
+        T0 : array-like of shape (n_samples, n_treatments) or scalar, optional
+            Baseline treatment level. If None, uses zeros.
+
+        T1 : array-like of shape (n_samples, n_treatments) or scalar, optional
+            Alternative treatment level. If None, uses ones.
+
+        Returns
+        -------
+        effects : ndarray of shape (n_samples,)
+            Estimated treatment effect for each sample.
+
+        Notes
+        -----
+        Based on counterfactual decomposition (Lemma 2.1 of paper).
+        """
+        check_is_fitted(self, ["coef_", "n_periods_"])
+        X = check_array(X)
+
+        n_treatments = self.coef_.shape[1]
+
+        # Handle T0, T1
+        if T0 is None:
+            T0 = np.zeros((X.shape[0], n_treatments))
+        elif np.isscalar(T0):
+            T0 = np.full((X.shape[0], n_treatments), T0)
+        else:
+            T0 = np.asarray(T0)
+            if T0.ndim == 1:
+                T0 = T0.reshape(-1, 1)
+
+        if T1 is None:
+            T1 = np.ones((X.shape[0], n_treatments))
+        elif np.isscalar(T1):
+            T1 = np.full((X.shape[0], n_treatments), T1)
+        else:
+            T1 = np.asarray(T1)
+            if T1.ndim == 1:
+                T1 = T1.reshape(-1, 1)
+
+        # Compute effect
+        effects = np.zeros(X.shape[0])
+        for t in range(self.n_periods_):
+            effects += (T1 - T0) @ self.coef_[t]
+
+        return effects
+
+    def effect(self, X, T0=0, T1=1):
+        """
+        Convenience method for predict() with default contrasts.
+
+        Parameters
+        ----------
+        X : array-like
+            Features.
+        T0, T1 : scalar or array-like
+            Treatment contrasts.
+
+        Returns
+        -------
+        effects : ndarray
+            Treatment effects.
+        """
+        return self.predict(X, T0=T0, T1=T1)
+
+    def effect_interval(self, X, T0=0, T1=1, alpha=0.05):
```

Comment:
I think this is specific to using a linear model as the effect estimator? Another idea for computing effect interval in the generalized case could be to do bootstrapping. We can define a bootstrapping class that takes any of our prediction classes and does bootstrapping to generate the interval.